src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java

```diff
@@ -0,0 +1,392 @@
+   * @return The Range that matches the error type
+   */
+  protected String inErrorCodeRange(String errorCode) {
+    if (!errorCode.isEmpty()) {
+      Integer errorCodeInt = Integer.getInteger(errorCode);
+      if (errorCodeInt >= 200 && errorCodeInt <= 299) {
+        errorCode = "200-299";
+      }
+    }
+    return errorCode;
+  }
+
+  /**
+   * Retrieves the primary email of the user with the access_token and auth
+   * token.
+   *
+   * @param credentials The json containing an access_token
+   * @return The unique user ID
+   * @throws ServiceException If there are issues determining the primary address
+   */
+  @Override
+  protected String getPrimaryEmail(JsonNode credentials) throws ServiceException {
+    JsonNode json = null;
+    final String authToken = credentials.get("access_token").asText();
+    String url = FacebookConstants.DEBUG_TOKEN_URI;
+    String queryString;
+    try {
+      queryString = "?input_token=" + authToken + "&access_token="
+          + URLEncoder.encode(getAppToken(), OAuth2Constants.ENCODING);
+    } catch (UnsupportedEncodingException e1) {
+      throw ServiceException.PARSE_ERROR("Url encoding the Facebook app token failed.", e1);
+    }
+
+    try {
+      final GetMethod request = new GetMethod(url);
+      request.setQueryString(queryString);
+      json = executeRequestForJson(request);
+    } catch (final IOException e) {
+      ZimbraLog.extensions.debug("There was an issue acquiring the authorization token.", e);
+      throw ServiceException.PROXY_ERROR("There was an issue acquiring an authorization token "
+          + "for this user.", null);
+    }
+    String userId = json.get("data").get("user_id").asText();
+    if (userId.isEmpty()) {
+      throw ServiceException.NOT_FOUND("User Id not found in json data. " + json.toString());
+    }
+
+    return userId;
+  }
+
+  /**
+   * Retrieves the App Token.
+   *
+   * @return The Facebook App token
+   * @throws ServiceException If there was an issue with the request
+   * @throws GenericOAuthException If there are issues acquiring the user id
+   */
+  protected String getAppToken() throws ServiceException {
+    JsonNode json = null;
+    String url = FacebookConstants.AUTHENTICATE_URI;
+    String queryString = "?client_id=" + this.clientId + "&client_secret="
+        + this.clientSecret + "&grant_type=client_credentials";
+    try {
+      final GetMethod request = new GetMethod(url);
+      request.setQueryString(queryString);
+      json = executeRequestForJson(request);
+    } catch (final IOException e) {
+      ZimbraLog.extensions.debug("There was an issue acquiring the app access token.", e);
+      throw ServiceException.PROXY_ERROR("There was an issue acquiring the Facebook app "
+        + "access token.", null);
+    }
+
+    String appAccessToken = json.get("access_token").asText();
+    if (appAccessToken.isEmpty()) {
+      throw ServiceException.NOT_FOUND("The Facebook App token not found in "
```

Comment:
Use a similar message as the other implementations do for primary email - see Yahoo's example.